sycl/plugins/unified_runtime/ur/adapters/hip/context.cpp

```diff
@@ -97,40 +83,8 @@ urContextGetInfo(ur_context_handle_t hContext, ur_context_info_t propName,
 
 UR_APIEXPORT ur_result_t UR_APICALL
 urContextRelease(ur_context_handle_t hContext) {
-  if (hContext->decrementReferenceCount() > 0) {
-    return UR_RESULT_SUCCESS;
-  }
-  hContext->invokeExtendedDeleters();
-
-  std::unique_ptr<ur_context_handle_t_> context{hContext};
-
-  if (!hContext->isPrimary()) {
-    hipCtx_t HIPCtxt = hContext->get();
-    // hipCtxSynchronize is not supported for AMD platform so we can just
-    // destroy the context, for NVIDIA make sure it's synchronized.
-#if defined(__HIP_PLATFORM_NVIDIA__)
-    hipCtx_t Current = nullptr;
-    UR_CHECK_ERROR(hipCtxGetCurrent(&Current));
-    if (HIPCtxt != Current) {
-      UR_CHECK_ERROR(hipCtxPushCurrent(HIPCtxt));
-    }
-    UR_CHECK_ERROR(hipCtxSynchronize());
-    UR_CHECK_ERROR(hipCtxGetCurrent(&Current));
-    if (HIPCtxt == Current) {
-      UR_CHECK_ERROR(hipCtxPopCurrent(&Current));
-    }
-#endif
-    return UR_CHECK_ERROR(hipCtxDestroy(HIPCtxt));
-  } else {
-    // Primary context is not destroyed, but released
-    hipDevice_t HIPDev = hContext->getDevice()->get();
-    hipCtx_t Current;
-    UR_CHECK_ERROR(hipCtxPopCurrent(&Current));
-    return UR_CHECK_ERROR(hipDevicePrimaryCtxRelease(HIPDev));
-  }
-
-  hipCtx_t HIPCtxt = hContext->get();
-  return UR_CHECK_ERROR(hipCtxDestroy(HIPCtxt));
+  hContext->decrementReferenceCount();
```

Comment:
Is the intention to only reference count here and use `hipDevicePrimaryCtxRelease` only when a device is being destroyed?

Reply:
Yeah since the `ur_context_handle_t` will feasibly contain multiple devices (meaning multiple primary contexts) there is no longer a direct link to the `hipCtx_t`

Reply:
OK, do we need to sanity check the result of the decrement, to assert it never gets < 0, or is that not possible?

Reply:
I think it should be OK to be honest. One thing I'm not entirely sure about though - it seems to be on the user to delete the phContext. This could be an issue since there are no ur entry points that tell you the refcount of the `ur_context_handle_t`.

Or should `urContextRelease` call `delete this` once the refcount is zero?

Reply:
The way it's implemented now is that it will be deleted upon a call to `~ur_device_handle_t_()`, I don't think it's a problem, the only situation when it could be unclear would be if you decremented the count on the context to 0 without it being destroyed and then try to re-use it by retaining it again, but that is guarded against in: https://github.com/intel/llvm/pull/10514/files/a83109cf0021cca0e579abbaf4f7542502f9ec0a#diff-606e07ace1cc7d68983d97c70c9d4c5cc91ffd8277330bd3fb14f365aecb9f6bR92

Reply:
The native context is being released in `~ur_device_handle_t`. It turns out that we should be calling `delete` on the `hContext` if the refcount goes to zero. Change made